lib/semantics/check-nullify.cc

```diff
@@ -19,37 +19,28 @@
 #include "../parser/message.h"
 #include "../parser/parse-tree.h"
 
-#include "../parser/dump-parse-tree.h"
-#include <iostream>
-
 namespace Fortran::semantics {
 
 void NullifyChecker::Leave(const parser::NullifyStmt &nullifyStmt) {
-  // R938
   for (const parser::PointerObject &pointerObject : nullifyStmt.v) {
-    // R939
     std::visit(
         common::visitors{
             [&](const parser::Name &name) {
               auto const *symbol{name.symbol};
               if (!IsVariableName(*symbol) && !IsProcName(*symbol)) {
                 context_.messages().Say(name.source,
-                    "name must be a variable or procedure pointer name"_err_en_US);
+                    "name in NULLIFY statement must be a variable or procedure pointer name"_err_en_US);
               } else if (!IsPointer(*symbol)) {  // C951
                 context_.messages().Say(name.source,
-                    "name must have the POINTER attribute"_err_en_US);
+                    "name in NULLIFY statement must have the POINTER attribute"_err_en_US);
               }
             },
             [&](const parser::StructureComponent &structureComponent) {
               evaluate::ExpressionAnalyzer analyzer{context_};
               if (MaybeExpr checked{analyzer.Analyze(structureComponent)}) {
```

Comment:
I don't think the call to Analyze is needed either. The symbol should already be filled in by name resolution or we won't get this far.

Reply:
The structure component might be complex, like a(i + 4)%b -- I think semantics needs to be called to check the "i + 4" for example, or that the index expression is an integer expression.  Does that happen in name resolution?

Reply:
Expression analysis (in ExprChecker) should analyze all of the top-level Exprs. That would include `i+4` in `nullify(a(i+4)%b)` because it is not a sub-expression of a bigger expression.

But I'm not sure about when `a(i+4)` gets analyzed because it's a `DataRef` that's not part of an `Expr`.

Reply:
Right now, the StructureComponents in NULLIFY are not handled during the expression analysis phase.  Strictly speaking, it's not an Expr but we can decide to extend the top-level expression analysis to handle this case (and others, e.g. AllocateObject).  Let's talk to Peter about it.

Reply:
It might be best to extend the grammar to accept an expression here, wrapped in a new constraint check.  Then the expression gets analyzed automatically, its representation gets saved for the CFG, and syntax errors get better recovery.

Reply:
+1